src/Processors/QueryPlan/JoinStepLogical.h

```diff
@@ -44,13 +53,25 @@ class JoinStepLogical final : public IQueryPlanStep
     JoinStepLogical(
         SharedHeader left_header_,
         SharedHeader right_header_,
-        JoinInfo join_info_,
+        JoinOperator join_operator_,
         JoinExpressionActions join_expression_actions_,
-        Names required_output_columns_,
+        const NameSet & required_output_columns_,
+        const std::unordered_map<String, const ActionsDAG::Node *> & changed_types,
```

Comment:
Let's add a comment on what `changed_types` is.

Now, it's only needed for `using` columns, to cast those into supertype.
E.g. for `JOIN ... using (x)` the `join_expression_actions` dag may contain chain like `INPUT x -> CAST(x, supertype) -> ALIAS x`; then `changed_types` will contain `"x" -> const Node * ALIAS x`